src/kOS/AddOns/AddonManager.cs

```diff
@@ -0,0 +1,41 @@
+﻿using kOS.Safe.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace kOS.AddOns
+{
+    [AssemblyWalk(AttributeType = typeof(kOSAddonAttribute), InherritedType = typeof(Suffixed.Addon), StaticRegisterMethod = "RegisterMethod")]
+    public class AddonManager
+    {
+        public Dictionary<string, Suffixed.Addon> AllAddons { get; private set; }
+        private static readonly Dictionary<kOSAddonAttribute, Type> rawAttributes = new Dictionary<kOSAddonAttribute, Type>();
+        private readonly SharedObjects shared;
+
+        public AddonManager(SharedObjects sharedObj)
+        {
+            AllAddons = new Dictionary<string, Suffixed.Addon>(StringComparer.OrdinalIgnoreCase);
+            foreach (kOSAddonAttribute attr in rawAttributes.Keys)
+            {
+                AllAddons.Add(attr.Identifier, (Suffixed.Addon)Activator.CreateInstance(rawAttributes[attr], sharedObj));
+            }
```

Comment:
It's not that I didn't understand the order of things.  It's that I didn't notice that it was static (a prerequisite for it being possible for it to be populated prior to the constructor getting called.)

So this seems strange to me.  If you have two different instances of `AddonManager`, then they both share the same static instance of `rawAttributes` and both try to run this code.

If the intention was that there must only ever be one instance of AddonManager (so that scenario of having two of them I described doesn't happen), then it looks like an odd archetecture to store half the one-instance information in static members and store the other half of the one-instance information in dynamic members.